relay-server/src/actors/project.rs

```diff
@@ -339,19 +340,14 @@ impl ProjectState {
                 return Err(DiscardReason::ProjectId);
             }
 
-            // since the config has been fetched recently, we assume unknown
-            // public keys do not exist and drop events eagerly. proxy mode is
-            // an exception, where public keys are backfilled lazily after
-            // events are sent to the upstream.
-            match self.get_public_key_status(meta.public_key()) {
-                PublicKeyStatus::Enabled => Ok(()),
-                PublicKeyStatus::Disabled => Err(DiscardReason::ProjectId),
-                PublicKeyStatus::Unknown => match config.relay_mode() {
-                    RelayMode::Proxy => Ok(()),
-                    _ => Err(DiscardReason::ProjectId),
-                },
+            // sanity-check that the up-to-date state has a matching public key loaded.
+            if !self.is_matching_key(meta.public_key()) {
```

Comment:
I believe that this is potentially no longer what we want. The old code only checked for disabled project states if the state is not hard outdated. If the state is cached as disabled but outdated, then we proceed to update it and then check again. We were generally quite lenient with this to allow proxies to forward events even if they don't yet know about a certain project (NOTE: A "missing" project is reported as `disabled: true`). 

However, the code also stated this:

> we assume it is unlikely to re-activate a disabled public key.	
> thus we handle events pretending the config is still valid,	
> except queueing events for unknown DSNs as they might have become	
> available in the meanwhile.	

This means that we **did** check for disabled DSNs even if the state was hard outdated. This is to distinguish disabling a project key vs not yet knowing about a project. 

@untitaker I think we have to re-introduce the disabled DSN state somehow. Particularly: 
- If a project is missing, then report it as `disabled: true`. 
- If a key is explicitly disabled, then report it without a key, or with a disabled key
- Always reject if a key is explicitly disabled.
- If the state is hard outdated, then skip the validation of `project.disabled()`.
- If the state is up-to-date, then validate `disabled()`.

Alternatively, we could indicate an explicit state for disabled vs missing.

Reply:
Okay, I updated this now according to the above logic. To reiterate:

- Missing projects, disabled projects, missing keys, and disabled keys are all reported and stored as `disabled: True`. 
- We only validate the `disabled` flag as long as the state is up-to-date. This includes the grace period.
- If the state is hard outdated, we buffer the event and then fetch from the upstream. 

This has two differences to the previous behavior:
1. Even on hard outdated states, Relays would reject events based on disabled project keys in the fast-path. This no longer happens, instead we buffer until we can update the state.
2. Proxy relays will now reject events if they have an up-to-date negative cache for a key. Previously if they knew the project exists and is enabled, but they don't know about a new key, they would just forward. We can no longer do this, because the new state does not allow to differentiate between disabled projects and disabled keys. I think this is a fair tradeoff.